# FILE where original DATA is analyzed and Transformed 


Our data is an extract of what we might find in our Google Analytics Data Stored in our Database

----------
## Analysis of the Data 

In [82]:
import pandas as pd

df = pd.read_csv("google data/data.csv")
clicks_df = df.loc[df.Click == True]
buy_df = pd.read_csv("google data/purchases.csv")

In [83]:
clicks_df.head()  # df where is located user liking behavior linked to clicked products

visitNumber     visitId  visitStartTime      date  total_visits  \
0            2  1470857552      1470857552  20160810             1   
1            1  1470862136      1470862136  20160810             1   
2            1  1470848713      1470848713  20160810             1   
3            4  1470815734      1470815734  20160810             1   
4            4  1470815734      1470815734  20160810             1   

   total_hits  total_pageviews  total_timeOnSite  total_bounces  \
0           5                4             125.0            NaN   
1           5                4             341.0            NaN   
2           6                5             106.0            NaN   
3           6                3               8.0            NaN   
4           6                3               8.0            NaN   

   total_newVisits  ...                            Category    Variant  \
0              NaN  ...           Home/Limited Supply/Bags/  (not set)   
1              1.0  ...           Home/Limited Supply/Bags/  (not set)   
2              1.0  ...  Home/Apparel/Men's/Men's-T-Shirts/  (not set)   
3              NaN  ...                        Home/Office/  (not set)   
4              NaN  ...                        Home/Office/  (not set)   

   ProductList  transactionId Revenue itemTransactionId Click TotalPaid  \
0     Category            NaN     NaN               NaN  True       0.0   
1     Category            NaN     NaN               NaN  True       0.0   
2     Category            NaN     NaN               NaN  True       0.0   
3     Category            NaN     NaN               NaN  True       0.0   
4     Category            NaN     NaN               NaN  True       0.0   

   Price Quantity  
0  99.99      NaN  
1   8.79      NaN  
2  16.99      NaN  
3   0.99      NaN  
4   0.99      NaN  

[5 rows x 34 columns]

In [4]:
buy_df.head()

visitNumber     visitId  visitStartTime      date  total_visits  \
0            2  1472571985      1472571985  20160830             1   
1            1  1472606991      1472606991  20160830             1   
2            3  1472599227      1472599227  20160830             1   
3            1  1472594459      1472594459  20160830             1   
4            3  1472604143      1472604143  20160830             1   

   total_hits  total_pageviews  total_timeOnSite  total_bounces  \
0          13               11               305            NaN   
1          13               12               392            NaN   
2          14               12               192            NaN   
3          16               14               685            NaN   
4          16               13               895            NaN   

   total_newVisits  ...  itemTransactionId  \
0              NaN  ...    ORD201608301467   
1              1.0  ...    ORD201608301495   
2              NaN  ...    ORD201608301510   
3              1.0  ...    ORD201608301546   
4              NaN  ...    ORD201608301559   

                                ProductName      ProductSKU   Category  \
0      24 oz YouTube Sergeant Stripe Bottle  GGOEYDHJ019399  (not set)   
1     Google Laptop and Cell Phone Stickers  GGOEGFKQ020399  (not set)   
2   20 oz Stainless Steel Insulated Tumbler  GGOEGDHQ014899  (not set)   
3           Google Stretch Fit Hat M/L Navy  GGOEGHPL003214  (not set)   
4  Google 17oz Stainless Steel Sport Bottle  GGOEGDHC074099  (not set)   

              Variant ProductList  Revenue   TotalCost   Price  Quantity  
0  Single Option Only   (not set)    20.91        7.99    7.99         1  
1  Single Option Only   (not set)   112.00       99.50    1.99        50  
2  Single Option Only   (not set)    11.24       19.99   19.99         1  
3                 M/L   (not set)    29.49       21.99   21.99         1  
4  Single Option Only   (not set)    33.89       18.99   18.99         1  

[5 rows x 32 columns]

As we can see, the dataframe contains a description of all sifferent features that we can use for creating a valid analysis and further (the recommendation system). The most important part is that this data is obtained from a Google's BigQuery use case in this [Kaggle](https://www.kaggle.com/code/paultimothymooney/how-to-query-the-google-analytics-sample-dataset/notebook?scriptVersionId=5165120)

The most important features were selected from a list of features. I used a query to retrieve the products that were clicked by the users, as well as the confirmed purchases. This, along side some behavioral features like the time on which the users entered the site, the page views, time on site and other, all this can allow us to create behavior clusters in which we can create a user classification system and later classify behaviors into this already formed groups.



In [5]:
buy_df.drop_duplicates(
    subset="transactionId", keep="first", inplace=True
)  # Drop duplicate orders

In [6]:
buy_df["ProductName"].value_counts()  # Products that have being purchased before

Google Men's 100% Cotton Short Sleeve Hero Tee Navy     30
Google Men's  Zip Hoodie                                30
Google Men's 100% Cotton Short Sleeve Hero Tee Black    27
Google Men's 100% Cotton Short Sleeve Hero Tee White    25
Engraved Ceramic Google Mug                             23
                                                        ..
Google Executive Umbrella                                1
Google Accent Insulated Stainless Steel Bottle           1
Gel Roller Pen                                           1
Google Men's Performance Tee Gunmetal                    1
Bic Digital Clic Stic Stylus Pen                         1
Name: ProductName, Length: 248, dtype: int64

In [7]:
clicks_df["ProductName"].value_counts()  # Products that have being clicked

Mistral Rucksack                                          633
Deluge Waterproof Backpack                                575
Oasis Backpack                                            553
Keyboard DOT Sticker                                      552
Google Men's 100% Cotton Short Sleeve Hero Tee White      466
                                                         ... 
Google Toddler Tee Fruit Games Cherries                     3
Google Men's Short Sleeve Performance Badge Tee Pewter      2
Gift Card - $250.00                                         1
Google Toddler Tee Fruit Games Lemon                        1
Google Youth Tee Fruit Games Pineapple                      1
Name: ProductName, Length: 266, dtype: int64

We have to see how variables like Hits and Time on Site are distributed.

In [8]:
features_list = [
    "visitNumber",
    "total_visits",
    "total_hits",
    "Price",
    "total_timeOnSite",
    "total_newVisits",
    "total_pageviews",
]


clicks_df.loc[:, features_list].describe().round(
    2
)  # This would be the behavioral features that we have available

visitNumber  total_visits  total_hits     Price  total_timeOnSite  \
count     20878.00       20878.0    20878.00  20878.00          20863.00   
mean          4.22           1.0       50.96     28.11            923.26   
std          13.46           0.0       51.92     30.15           1060.52   
min           1.00           1.0        2.00      0.79              1.00   
25%           1.00           1.0       17.00      5.99            243.00   
50%           1.00           1.0       32.00     16.99            547.00   
75%           3.00           1.0       65.00     34.99           1210.00   
max         156.00           1.0      301.00    250.00           7508.00   

       total_newVisits  total_pageviews  
count          11736.0         20878.00  
mean               1.0            33.45  
std                0.0            32.42  
min                1.0             1.00  
25%                1.0            12.00  
50%                1.0            22.00  
75%                1.0            43.00  
max                1.0           188.00

---------
## Embeddings 



Our system will work with what is called embedding a text in NLP. We take our text, map it into tokens, which are basically small id's for the word. Sometimes words are divided in multiple tokens. Which account for the fact that some words that are very similar or share a common root can be represented similarly. Those tokens are unique id's to the realm of what will be our embeddings realm. Which is nothing more than a vectorized space in which this tokens are represented in a numeric form --> $ \text{token:} i \rightarrow \vec{w_i} = <0.423,3.141,.674, ..., n>$.

When we have a sentence or a paragraph, embeddings are usually the average of all the tokens that we find in that structure, and there we have it, our embedding system. Usually this embeddings spaces are then trained on a corpus of text to represent different features. This is similar to what our brain does with sentences: we remember the nouns, the verbs, the grammar rules, and other complex rules that make our understanding of the language as it is.

On this demo, we use ***multilingual-e5-base*** model for creating a 768 size vector that will account for the features in our product title, which is the main thing we want to use as a driver of recommendations.


In [9]:
import sys

sys.path.append("..")

from main.models import ModelClass

model = ModelClass()

In [8]:
model.create_embeddings("Hello world")  # Use method create embeddings to use it.

array([[ 3.35431285e-02,  4.46086712e-02, -4.48141305e-04,
        -3.62625433e-04,  2.88973264e-02, -3.31031904e-02,
        -1.90947279e-02, -3.67002678e-03,  3.32840942e-02,
         3.20200399e-02, -2.73499619e-02, -1.29014198e-02,
         1.63098887e-01,  3.99894156e-02, -3.45210619e-02,
        -2.77194325e-02,  1.10267121e-02, -2.44755037e-02,
         2.67019086e-02, -1.29492851e-02,  5.62458374e-02,
        -2.56671403e-02,  3.23554426e-02, -3.01293898e-02,
         2.17710696e-02, -9.27064288e-03,  2.39941813e-02,
         1.11178188e-02, -3.06521188e-02,  3.21589038e-02,
         9.67866369e-03, -7.68294372e-03,  3.73637117e-02,
         1.61362011e-02,  3.65321301e-02,  4.55307253e-02,
        -8.69155489e-03, -3.23016718e-02,  4.63906303e-02,
         2.58311220e-02,  1.76636688e-02,  3.31863835e-02,
        -2.47860118e-03, -4.25862223e-02,  2.99789384e-02,
         2.67586531e-03,  1.81598496e-02,  1.37071135e-02,
        -3.10147200e-02, -3.98573913e-02,  2.17524301e-0

Then, we would either need to run this over all the dataframes rows, which is not wrong, but will take some time that we need ;)

Therefore we are going to create the embeddings in a separate file, associated with a title. Since there are onlt 266 products, we dont need more than that. But now is the part where we make our first decision. Either we do some feature engineering for creating a new df which contains unique visitID clicks, append all the products clicked in just one big string, and then perform the embeddings process, or, we can create the embeddings for the 266 products and then create some average. The first process seems more accurate but will take longer, and the second one is going to take less but could be less straightforward. 

We are going then to take both approaches to see where both go.


### First Approach - Concatenate Likes and Embed the  "Like List"

In [10]:
# Group by 'visitId' and aggregate 'Product name' using text join
grouped_df = (
    df.groupby("visitId")["ProductName"].apply(lambda x: ", ".join(x)).reset_index()
)

# Rename the aggregated column
grouped_df.rename(columns={"ProductName": "Clicks"}, inplace=True)

print(grouped_df)

         visitId                                             Clicks
0     1470035161                                  Google Power Bank
1     1470035457                         Deluge Waterproof Backpack
2     1470035521  Seat Pack Organizer, Seat Pack Organizer, Seat...
3     1470035892  Google Men's 100% Cotton Short Sleeve Hero Tee...
4     1470036291  Google Men's 100% Cotton Short Sleeve Hero Tee...
...          ...                                                ...
4630  1470896943  Google Toddler Tee Fruit Games Pineapple, Goog...
4631  1470896966  Google Canvas Tote Natural/Navy, Mistral Rucks...
4632  1470897377      Google Women's Lightweight Microfleece Jacket
4633  1470897905  Google Men's 100% Cotton Short Sleeve Hero Tee...
4634  1470897937          Google Women's Performance Golf Polo Blue

[4635 rows x 2 columns]


In [ ]:
from tqdm.notebook import tqdm_notebook

tqdm_notebook.pandas()  # fancy progress Bar

grouped_df["Embedings"] = grouped_df.Clicks.progress_apply(
    lambda x: model.create_embeddings(x)
)  # Make the embeddings for the Clicks column

In [54]:
grouped_df.Embeddings  # The embeddings

0       [[0.021230968, 0.06723951, -0.03646528, -0.005...
1       [[0.021318657, 0.04240716, -0.007994821, 0.004...
2       [[0.002186621, 0.051950835, -0.004776678, 0.02...
3       [[0.00616312, 0.04824848, -0.035558887, 0.0190...
4       [[-9.8550074e-05, 0.034511797, -0.016930008, 0...
                              ...                        
4630    [[0.0056050452, 0.029619096, -0.017937213, 0.0...
4631    [[0.02744697, 0.05886887, -0.022487251, 0.0082...
4632    [[0.010117159, 0.042951882, -0.020193057, -0.0...
4633    [[0.0072451867, 0.036614206, -0.039741036, 0.0...
4634    [[0.02905081, 0.046681486, -0.016705029, 0.021...
Name: Embeddings, Length: 4635, dtype: object

### Second Approach - Create our Embeddings Master File 

In [ ]:
tqdm_notebook.pandas()

inv_df = pd.DataFrame(
    [clicks_df.ProductName.unique(), clicks_df["ProductName"].value_counts()]
).T  # Products that have being clicked
inv_df.rename({0: "Product", 1: "Count"}, axis=1, inplace=True)

inv_df["Embeddings"] = inv_df.Product.progress_apply(model.create_embeddings)

In [59]:
inv_df.Embeddings

0      [[0.014249629, 0.023254616, -0.006114072, 0.00...
1      [[0.03786302, 0.03239363, -0.00020421806, 0.02...
2      [[0.00114997, 0.03650656, -0.035562657, 0.0202...
3      [[0.028030083, 0.038984273, -0.02813911, 0.034...
4      [[0.015068452, 0.05567977, -0.016779918, 0.011...
                             ...                        
261    [[0.021007422, 0.04520183, -0.023061385, 0.041...
262    [[0.013173798, 0.00889726, -0.012541152, 0.015...
263    [[0.008392215, 0.055684857, -0.011121839, -0.0...
264    [[-0.01466453, 0.04144204, -0.04656081, 0.0284...
265    [[0.0017286192, 0.04235564, -0.026112005, 0.01...
Name: Embeddings, Length: 266, dtype: object

In [17]:
## Now use this inv_df and original df to make a new df containing the embeddings.

clicks_df.rename({"ProductName": "Product"}, axis=1, inplace=True)

clicks_df = pd.merge(
    inv_df[["Product", "Embeddings"]], clicks_df, on="Product", how="right")

clicks_df

Product  \
0                                       Oasis Backpack   
1                         Set of 3 Nested Travel Cases   
2            YouTube Men's Short Sleeve Hero Tee Black   
3                                             Maze Pen   
4                                             Maze Pen   
...                                                ...   
20873             Google Women's Hero V-Neck Tee White   
20874             Google Women's Hero V-Neck Tee White   
20875       Google Women's Short Sleeve Hero Tee White   
20876  Google Women's Short Sleeve V-Neck Tee Lavender   
20877                     Google Women's Fleece Hoodie   

                                            Embeddings_x  \
0      [[0.014249629, 0.023254616, -0.006114072, 0.00...   
1      [[0.03786302, 0.03239363, -0.00020421806, 0.02...   
2      [[0.00114997, 0.03650656, -0.035562657, 0.0202...   
3      [[0.028030083, 0.038984273, -0.02813911, 0.034...   
4      [[0.028030083, 0.038984273, -0.02813911, 0.034...   
...                                                  ...   
20873  [[0.018843824, 0.053733025, -0.02561408, 0.008...   
20874  [[0.018843824, 0.053733025, -0.02561408, 0.008...   
20875  [[-5.1616345e-05, 0.047178693, -0.030958962, 0...   
20876  [[0.01456775, 0.050580125, -0.034778133, 0.027...   
20877  [[0.01493587, 0.05140108, -0.021872329, 0.0094...   

                                            Embeddings_y  visitNumber  \
0      [[0.014249629, 0.023254616, -0.006114072, 0.00...            2   
1      [[0.03786302, 0.03239363, -0.00020421806, 0.02...            1   
2      [[0.00114997, 0.03650656, -0.035562657, 0.0202...            1   
3      [[0.028030083, 0.038984273, -0.02813911, 0.034...            4   
4      [[0.028030083, 0.038984273, -0.02813911, 0.034...            4   
...                                                  ...          ...   
20873  [[0.018843824, 0.053733025, -0.02561408, 0.008...            7   
20874  [[0.018843824, 0.053733025, -0.02561408, 0.008...            7   
20875  [[-5.1616345e-05, 0.047178693, -0.030958962, 0...            7   
20876  [[0.01456775, 0.050580125, -0.034778133, 0.027...            7   
20877  [[0.01493587, 0.05140108, -0.021872329, 0.0094...            7   

          visitId  visitStartTime      date  total_visits  total_hits  \
0      1470857552      1470857552  20160810             1           5   
1      1470862136      1470862136  20160810             1           5   
2      1470848713      1470848713  20160810             1           6   
3      1470815734      1470815734  20160810             1           6   
4      1470815734      1470815734  20160810             1           6   
...           ...             ...       ...           ...         ...   
20873  1470120798      1470120798  20160801             1          34   
20874  1470120798      1470120798  20160801             1          34   
20875  1470120798      1470120798  20160801             1          34   
20876  1470120798      1470120798  20160801             1          34   
20877  1470120798      1470120798  20160801             1          34   

       total_pageviews  ...                                 Category  \
0                    4  ...                Home/Limited Supply/Bags/   
1                    4  ...                Home/Limited Supply/Bags/   
2                    5  ...       Home/Apparel/Men's/Men's-T-Shirts/   
3                    3  ...                             Home/Office/   
4                    3  ...                             Home/Office/   
...                ...  ...                                      ...   
20873               18  ...   Home/Apparel/Women's/Women's-T-Shirts/   
20874               18  ...   Home/Apparel/Women's/Women's-T-Shirts/   
20875               18  ...   Home/Apparel/Women's/Women's-T-Shirts/   
20876               18  ...   Home/Apparel/Women's/Women's-T-Shirts/   
20877               18  ...  Home/Apparel/Women's/Women's-Outerwear/   

         V

In [ ]:
##  Now we store pur embeddings so that we dont do the same all over again

import os

inv_df.to_hdf("google data/embeddings/inventory_embeddings.h5", key="data", mode="w") 
grouped_df.to_hdf("google data/embeddings/likes_embeddings.h5", key="data", mode="w")

-------
## Second phase of the walkthrough here with the embeddings completed.


We can start directly her eif the embeddings is not in what we are interested in.

Now we need to implement something that creates a scoring based on the behavioral features. This way, the recommendation process will have more sense. In this use case, we are going to use the original to extract a scoring function using a "Sell" as the target variable that we want to predict. This is the general objective of the e-commerce use case. You may think of how are we going to create a score based on this. 

The answer comes as a trick in transforming the data. We are going to take one set of data that contains visit ID who end up in purchases and cases in which are not. Then using a classic sigmoid head at the end of whatever our model is going to be, we end up getting the probabilities of one visit ID ending up in a purchase, which turns this also into a binary classification problem. Our score is going to be the probability that ends up in a selling.

### Data Preparation


In [223]:
import pandas as pd 
import numpy as np

emb = pd.read_hdf('google data/embeddings/inventory_embeddings.h5',key='data') # Embeddings of the inventory. 

buy_df = pd.read_csv("google data/purchases.csv") # only purchases 
data = pd.read_csv("google data/data.csv")        # mixed df
data = data.loc[data.Click == True]

data.rename({"ProductName": "Product"}, axis=1, inplace=True)
buy_df.rename({"ProductName": "Product"}, axis=1, inplace=True)

data = pd.merge(
    emb[["Product", "Embeddings"]], data, on="Product", how="right") # Add the embeddings Column

buy_df = pd.merge( 
    emb[["Product", "Embeddings"]], buy_df, on="Product", how="right") # Add the embeddings Column


data.dropna(subset='Embeddings',inplace=True)
buy_df.dropna(subset='Embeddings',inplace=True)


features_list = [
    "visitId",
    "visitNumber",
    "total_visits",
    "total_hits",
    "Price",
    "total_timeOnSite",
    "total_newVisits",
    "total_pageviews",
]

## TRANSFORMATIONS -------------------

# Group by 'visitId' and aggregate 'Product name' using text join

data_x =  buy_df.groupby("visitId")["Product"].apply(lambda x: ", ".join(x)).reset_index()
data_xx = data.groupby("visitId")["Product"].apply(lambda x: ", ".join(x)).reset_index()

# Group by visitid and aggregate 'Revenue'

data_y = buy_df.groupby("visitId")['Revenue'].apply(lambda x: 1 if np.sum(x)>0 else 0).reset_index()
data_yy =  data.groupby("visitId")['Revenue'].apply(lambda x: 1 if np.sum(x)>0 else 0).reset_index()

# Combine the embeddings of all the products bought or clicked in the same session

emb1 = buy_df.groupby("visitId")['Embeddings'].apply(lambda x: np.stack(x,axis=0)).reset_index()
emb2 = data.groupby("visitId")['Embeddings'].apply(lambda x: np.stack(x,axis=0)).reset_index()

# Rename the aggregated column

data_x.rename(columns={"ProductName": "Clicks"}, inplace=True)
data_y.rename(columns={"Revenue": "Sells"}, inplace=True)
data_xx.rename(columns={"ProductName": "Clicks"}, inplace=True)
data_yy.rename(columns={"Revenue": "Sells"}, inplace=True)

buy_df = (
    pd.merge(data_x, buy_df.loc[:, features_list], on="visitId", how="left")
    .drop_duplicates(subset="visitId")
    .reset_index(drop=True)
)

data = (
    pd.merge(data_xx, data.loc[:, features_list], on="visitId", how="left")
    .drop_duplicates(subset="visitId")
    .reset_index(drop=True)
)

buy_df.total_newVisits =  buy_df.total_newVisits.apply(lambda x: 0 if pd.isna(x) else 1)
data.total_newVisits =  data.total_newVisits.apply(lambda x: 0 if pd.isna(x) else 1)

buy_df = pd.merge(data_y, buy_df,on="visitId", how="right")  # DF with NOTHING MORE THAN COMPLETED ORDERS
data = pd.merge(data_yy, data,on="visitId", how="right")  # DF with NOTHING MORE THAN COMPLETED ORDERS

# Append the embeddings 
buy_df = pd.merge(emb1, buy_df,on="visitId", how="right")
data = pd.merge(emb2, data,on="visitId", how="right")


In [224]:
size = data.Embeddings.apply(lambda x: x.shape[0])
size_ = buy_df.Embeddings.apply(lambda x: x.shape[0])

print(size.describe(),"\n")  # We can see here that instead of just getting the mean between the things viewed or bought on each session. We could stack along a new dimension for working in a more complex way with them.
print(size_.describe())      # Since the distributions indicate that around 5 objects or items is the most common number, we can either trim or pad to adjust to that length

count    4614.000000
mean        4.524924
std         6.155768
min         1.000000
25%         1.000000
50%         3.000000
75%         5.000000
max       111.000000
Name: Embeddings, dtype: float64 

count    1016.000000
mean        3.375000
std         3.215131
min         1.000000
25%         1.000000
50%         2.000000
75%         4.000000
max        27.000000
Name: Embeddings, dtype: float64


In [225]:
print(f"the number of sessions in which one order was completed in the first data source is: {(data.Sells==1).sum()} from {data.shape[0]} instances." )
print(f"the number of sessions in which one order was completed in the second data source is: {(buy_df.Sells==1).sum()} from {buy_df.shape[0]} instances.")

the number of sessions in which one order was completed in the first data source is: 0 from 4614 instances.
the number of sessions in which one order was completed in the second data source is: 1016 from 1016 instances.


In [226]:
df = pd.concat([data,buy_df], axis=0,ignore_index=True)
df.drop_duplicates(subset='visitId',inplace=True)
df.dropna(inplace=True)
df  # New df with whole data.

visitId                                         Embeddings  Sells  \
0     1470035161  [[[0.021230968, 0.06723951, -0.03646528, -0.00...      0   
1     1470035457  [[[0.021318657, 0.04240716, -0.007994821, 0.00...      0   
2     1470035521  [[[0.015425235, 0.023942046, 0.0014200645, 0.0...      0   
3     1470035892  [[[0.00616312, 0.04824848, -0.035558887, 0.019...      0   
4     1470036291  [[[0.007224333, 0.037768986, -0.038783304, 0.0...      0   
...          ...                                                ...    ...   
5625  1472602179  [[[0.0111623695, 0.0331794, -0.040837016, 0.01...      1   
5626  1472604143  [[[0.01618643, 0.05558616, -0.027737834, -0.00...      1   
5627  1472606991  [[[0.011397464, 0.06847937, -0.0370449, -0.012...      1   
5628  1472610310  [[[0.0105100805, 0.07540334, -0.035388947, 0.0...      1   
5629  1472615797  [[[0.0072451867, 0.036614206, -0.039741036, 0....      1   

                                                Product  visitNumber  \
0                                     Google Power Bank            1   
1                            Deluge Waterproof Backpack            3   
2     Seat Pack Organizer, Seat Pack Organizer, Seat...            1   
3     Google Men's 100% Cotton Short Sleeve Hero Tee...            1   
4     Google Men's 100% Cotton Short Sleeve Hero Tee...            7   
...                                                 ...          ...   
5625  Google Men's 100% Cotton Short Sleeve Hero Tee...            1   
5626           Google 17oz Stainless Steel Sport Bottle            3   
5627              Google Laptop and Cell Phone Stickers            1   
5628  Google Canvas Tote Natural/Navy, Badge Holder,...            1   
5629  Google Men's 100% Cotton Short Sleeve Hero Tee...            3   

      total_visits  total_hits  Price  total_timeOnSite  total_newVisits  \
0                1           9  16.99              89.0                1   
1                1           7  99.99             421.0                0   
2                1          53  11.99             755.0                1   
3                1           8  16.99              82.0                1   
4                1          33  16.99             465.0                0   
...            ...         ...    ...               ...              ...   
5625             1          38  16.99             494.0                1   
5626             1          16  18.99             895.0                0   
5627             1          13   1.99             392.0                1   
5628             1          53  12.79            2430.0                1   
5629             1          23  13.59             568.0                0   

      total_pageviews  
0                   7  
1                   5  
2                  36  
3                   6  
4                  20  
...               ...  
5625               27  
5626               13  
5627               12  
5628               44  
5629               18  

[5381 rows x 11 columns]

In [227]:
inst = df.Sells.value_counts()
inst.index = ["Didn't Sell","Sell"]   # type: ignore
inst # Now we see we have a better distribution of out target variable.

Didn't Sell    4599
Sell            782
Name: Sells, dtype: int64

 Since the distribution is not very well done in terms of balance between sells and not sold, we need to take that into account when watching how the model goes

In [228]:

import random

# Filter the DataFrame to balance the target variable
df_balanced = df.groupby('Sells').apply(lambda x: x.sample(n=700) if x['Sells'].values[0] == 1 else x.sample(n=1500))
df_balanced = df_balanced.droplevel(0)  # Remove the grouping level

# Create a list of all row indexes
all_indexes = list(df_balanced.index)

inst = df_balanced.Sells.value_counts()
inst.index = ["Didn't Sell","Sell"]   # type: ignore
inst # Now we see we have a better distribution of out target variable.

Didn't Sell    1500
Sell            700
Name: Sells, dtype: int64

In [229]:
# Specify the proportion of data to include in the training set
train_proportion = 0.8  # 80% for training, 20% for testing

# Calculate the number of rows for training and testing
num_rows = len(df_balanced)
num_train = int(train_proportion * num_rows)
num_test = num_rows - num_train

# Randomly sample indexes for the training set
train_indexes = random.sample(all_indexes, num_train)

# Remove the selected training indexes to get testing indexes
test_indexes = list(set(all_indexes) - set(train_indexes))

# Create the training and testing DataFrames using the sampled indexes
train_df = df_balanced.loc[train_indexes]
test_df = df_balanced.loc[test_indexes]

----------
## Scoring Model

Now we set up a simple Model that contains Linear Layers for beahvior features, another branch for processing the embeddings, and a Sigmoid head. 

First, we inspect the posibility of getting it done with some backend suppot of our GPU,which can greatly augment the velocity of computation and tehrefore trainig.

In [9]:
# check availability of gpu in you system 
import torch 

if torch.cuda.is_available() and torch.backends.cuda.is_built(): device='cuda'; print("Cuda Available") 
elif torch.backends.mps.is_available() and torch.backends.mps.is_built(): device='mps'; print('MPS backend available')
else: device = 'cpu'; print('CPU is going to be used') 


MPS backend available


We initiate and ttest with some toy data observing that it does give us an output containing (batch_size,1) were each element is the probability of that visit Session, ending in a Sell.

In [263]:
import torch
import torch.nn as nn
from tqdm.notebook import tqdm_notebook


class CombinedModel(nn.Module):
    def __init__(
        self,
        conv_in_channels,
        num_behavioural_features,
        conv_out_channels,
        conv_kernel_size=3,
        stride=3,
        pool_size=2,
    ):
        super(CombinedModel, self).__init__()

        # Linear layers for behavioral features
        self.behavioural_layers = nn.Sequential(
            nn.Linear(num_behavioural_features, 64),
            nn.ReLU(),
            nn.Linear(64, 32),
            nn.ReLU(),
            nn.Linear(32, 16),
            nn.ReLU(),
        )

        # Convolutional layers for embeddings
        self.conv_layers = nn.Sequential(
            nn.Conv1d(
                conv_in_channels,
                conv_out_channels,
                kernel_size=conv_kernel_size,
                stride=stride,
            ),
            nn.ReLU(),
            nn.MaxPool1d(kernel_size=32),
        )

        # Linear layer for combining features

        self.output_size = int(
            conv_out_channels
            * np.floor(((768 - conv_kernel_size) / stride) + 1)
            / pool_size
        )
        self.final_linear = nn.Linear(16 + self.output_size, 1)
        self.sigmoid = nn.Sigmoid()

    def forward(self, behavioural_features, embeddings):
        behavioural_output = self.behavioural_layers(behavioural_features)
        # embeddings = embeddings.unsqueeze(2)  # Add a channel dimension for conv1d
        conv_output = self.conv_layers(embeddings)
        conv_output = torch.flatten(conv_output, start_dim=1)  # Flatten conv output

        combined_features = torch.cat((behavioural_output, conv_output), dim=1)
        final_linear = self.final_linear(combined_features)
        final_output = self.sigmoid(final_linear)  # Get probabilities

        return final_output

    def train_model(
            self,
            train_loader,
            num_epochs=5,
            learning_rate=1e-3,
            validation_loader=None,
            device=None,
        ):
        if device is None:
            device = torch.device(
                "cuda"
                if torch.cuda.is_available()
                else ("mps" if torch.backends.mps.is_available() else "cpu")
            )
        else:
            device = "cpu"
            self.to(device)

        criterion = nn.BCELoss()                                           # Binary Cross Entropy
        optimizer = torch.optim.Adam(self.parameters(), lr=learning_rate)  # Adam Optimizer
        
        loss_hist = {'train':[],
                     'val':[]}

        for epoch in range(num_epochs):
            self.train()
            total_loss = 0.0

            data_generator = train_loader
            
            print(f"Epoch {epoch+1}/{num_epochs}")

            for i, (behavioural_features, embeddings, labels) in enumerate(
                data_generator, start=1
            ):
                behavioural_features = behavioural_features.to(device)
                embeddings = embeddings.to(device)
                labels = labels.to(device)

                optimizer.zero_grad()
                outputs = model(behavioural_features, embeddings)
                loss = criterion(outputs, labels)
                loss.backward()
                optimizer.step()

                total_loss += loss.item()

                progress_percentage = i / len(data_generator)
                progress_bar_length = 100
                num_chars = int(progress_percentage * progress_bar_length)
                progress_bar = "[" + "-" * num_chars + " " * (progress_bar_length - num_chars) + "]"
                
                # Clear the previous console message and update with current loss
                print(
                f"\r\033[1mBatch {i}/{len(data_generator)} - Loss: {total_loss / i:.4f}\033[0m {progress_bar} {progress_percentage * 100:.2f}%",
                end="",
            )
            
            loss_hist['train'].append(total_loss) # Append loss to history 

            print("")  # Move to a new line after completing the epoch

            # Validation loop

            if validation_loader:

                val_loader = validation_loader

                self.eval()
                val_loss = 0.0
                with torch.no_grad():
                    for behavioural_features, embeddings, labels in val_loader:
                        behavioural_features = behavioural_features.to(device)
                        embeddings = embeddings.to(device)
                        labels = labels.to(device)

                        val_outputs = self(behavioural_features, embeddings)
                        val_loss += criterion(val_outputs, labels).item()

                val_loss /= len(val_loader)
                print(
                    f"Validation Loss (Epoch {epoch + 1}/{num_epochs}): {val_loss:.4f}"
                )

                loss_hist['val'].append(val_loss) # Append loss to history 


# Instantiate the model
embedding_dim = 768
num_behavioural_features = 10  #  number of behavioural features
num_object_embedded = 5  #  number of in channels to apply conv1d to
conv_out_channels = 16
conv_kernel_size = 3

model = CombinedModel(
    num_object_embedded,
    num_behavioural_features,
    conv_out_channels,
    conv_kernel_size,
    pool_size=32,
)

# Generate some sample inputs
behavioural_features = torch.randn(16, num_behavioural_features)  # Batch size 16
embeddings = torch.randn(16, num_object_embedded, embedding_dim)

# Pass inputs through the model
output = model(behavioural_features, embeddings)
print(output)
print("\n", output.shape)

tensor([[0.4613],
        [0.5280],
        [0.5231],
        [0.5409],
        [0.5728],
        [0.5265],
        [0.5236],
        [0.5076],
        [0.5533],
        [0.5267],
        [0.5178],
        [0.4965],
        [0.4675],
        [0.4831],
        [0.5437],
        [0.4834]], grad_fn=<SigmoidBackward0>)

 torch.Size([16, 1])


--------
## Training the Score Model

#### Trim Or Pad Embeddings 

In [231]:
def trim_or_pad(seq:torch.Tensor,max_length=5):
    if seq.shape[1] < max_length:
        # If sequence length is less than max_length, pad with zeros
        pad_size = max_length - seq.shape[1]
        padding = torch.zeros(seq.shape[0], pad_size, seq.shape[2])
        return torch.cat((seq, padding), dim=1)
    else:
        # If sequence length is greater than max_length, randomly select max_length elements
        indices = np.random.choice(range(max_length),max_length)
        return seq[:, indices, :]
    

In [50]:
print(f"For a vector whose examples are more than 5 we trim leaving the size from {list(torch.randn(16, 13, 500).shape)} --> {list(trim_or_pad(torch.randn(16, 13, 500)).shape)}")
print(f"On the other hand, for a vector whose examples are less than 5 we pad leaving the size from {list(torch.randn(16, 2, 500).shape)} --> {list(trim_or_pad(torch.randn(16, 2, 500)).shape)}")

For a vector whose examples are more than 5 we trim leaving the size from [16, 13, 500] --> [16, 5, 500]
On the other hand, for a vector whose examples are less than 5 we pad leaving the size from [16, 2, 500] --> [16, 5, 500]


#### Data Loader

In [245]:

class CustomDataGenerator:
      
    def __init__(self, data, labels, batch_size=64, shuffle=True):
        self.data = data
        self.labels = labels
        self.batch_size = batch_size
        self.shuffle = shuffle
        self.indices = np.arange(labels.shape[0])
        if shuffle:
            np.random.shuffle(self.indices)
        self.current_idx = 0

    def __iter__(self):
        self.current_idx = 0
        return self

    def __next__(self):
        if self.current_idx >= self.labels.shape[0]:
            raise StopIteration

        batch_indices = self.indices[self.current_idx:self.current_idx + self.batch_size]

        features_batch = self.data['features'][batch_indices]
        embeddings_batch = self.data['embeddings'][batch_indices]
        labels_batch = self.labels[batch_indices]

        self.current_idx += self.batch_size

        return features_batch,embeddings_batch, labels_batch

    def __len__(self):
        return self.labels.shape[0] // self.batch_size

drop_features_names = ['visitId','Product']

train = train_df.drop(drop_features_names,axis=1).reset_index(drop=True)
test = test_df.drop(drop_features_names,axis=1).reset_index(drop=True)

train_embeddings = train.Embeddings.apply(lambda x : trim_or_pad(
    torch.tensor(x ,dtype=torch.float32).permute(1,0,2).contiguous())) # type: ignore

test_embeddings  = test.Embeddings.apply(lambda x: trim_or_pad(
    torch.tensor(x,dtype=torch.float32).permute(1,0,2).contiguous())) # type: ignore

train_labels = torch.tensor(train.Sells.values,dtype=torch.float32).view(-1,1)
test_labels = torch.tensor(test.Sells.values,dtype=torch.float32).view(-1,1)

train = {'features': torch.tensor(train.drop(['Sells', 'Embeddings'], axis=1).values, dtype=torch.float32),
         'embeddings': torch.concat(train_embeddings.to_list(),dim=0)}

test = {'features': torch.tensor(test.drop(['Sells', 'Embeddings'], axis=1).values, dtype=torch.float32),
         'embeddings': torch.concat(test_embeddings.to_list(),dim=0)}


In [281]:
# Usage example
train_generator = CustomDataGenerator(data=train, labels=train_labels, batch_size=32, shuffle=True)
test_generator = CustomDataGenerator(data=test, labels=test_labels, batch_size=32, shuffle=True)

# Example of iterating through the data generator
for data_batch in train_generator:
    features, embeddings , labels  = data_batch
    print(f"The embeddings have a shape of --> {list(embeddings.shape)}")
    print(f"The features have a shape of -->{list(features.shape)}")
    print(f"The true value labels have a shape of --> {list(labels.shape)}")
    break 
        
# Instantiate the model
embedding_dim = embeddings.shape[2]
num_behavioural_features = features.shape[1]  #  number of behavioural features
num_object_embedded = embeddings.shape[1]        #  number of in channels to apply conv1d to 
conv_out_channels = 16
conv_kernel_size = 3

model = CombinedModel(num_object_embedded, num_behavioural_features, conv_out_channels, conv_kernel_size,pool_size=32)

print(f"The model's output has a shape --> {list(model(features,embeddings).shape)}")

The embeddings have a shape of --> [32, 5, 768]
The features have a shape of -->[32, 7]
The true value labels have a shape of --> [32, 1]
The model's output has a shape --> [32, 1]



Now that everything is apparently ok, we go and train for a few epochs, which will proceed to the training phase. 

### Training Phase

In [282]:
import warnings

warnings.filterwarnings('ignore')

model.train_model(train_generator, num_epochs=300,
            learning_rate=1e-3,
            validation_loader=test_generator,
            device='cpu')

Epoch 1/300
Batch 55/55 - Loss: 0.6346 [----------------------------------------------------------------------------------------------------] 100.00%
Validation Loss (Epoch 1/300): 0.6109
Epoch 2/300
Batch 55/55 - Loss: 0.5307 [----------------------------------------------------------------------------------------------------] 100.00%
Validation Loss (Epoch 2/300): 0.5930
Epoch 3/300
Batch 55/55 - Loss: 0.5089 [----------------------------------------------------------------------------------------------------] 100.00%
Validation Loss (Epoch 3/300): 0.5881
Epoch 4/300
Batch 55/55 - Loss: 0.4770 [----------------------------------------------------------------------------------------------------] 100.00%
Validation Loss (Epoch 4/300): 0.5646
Epoch 5/300
Batch 55/55 - Loss: 0.4684 [----------------------------------------------------------------------------------------------------] 100.00%
Validation Loss (Epoch 5/300): 0.6302
Epoch 6/300
Batch 55/55 - Loss: 0.4660 [--------------------

In [283]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

predictions = []
ground_truth = []

with torch.no_grad():
    for features,embeddings,labels in test_generator:
        outputs = model(features,embeddings)
        predicted_class = np.round(outputs.detach().numpy())
        predictions.extend(predicted_class)
        ground_truth.extend(labels.detach().numpy())

print(f"Accuracy: {accuracy_score(y_true=ground_truth,y_pred=predictions)}","\n")
print(f"Classification Report:\n{classification_report(y_true=ground_truth,y_pred=predictions)}","\n")
print(f"Confusion Matrix:\n{confusion_matrix(y_true=ground_truth,y_pred=predictions)}")


Accuracy: 0.8113636363636364 

Classification Report:
              precision    recall  f1-score   support

         0.0       0.82      0.92      0.87       291
         1.0       0.79      0.60      0.68       149

    accuracy                           0.81       440
   macro avg       0.80      0.76      0.77       440
weighted avg       0.81      0.81      0.80       440
 

Confusion Matrix:
[[267  24]
 [ 59  90]]
